SyncManager: Optimise and fixes

gbjk opened the pull request:
This is a fairly invasive change which addresses the amount of work the sync manager does each cycle and the cycle intervals. We switch to using discrete locks for each type of work on each pair, so each worker can take a discrete chunk of work safely. For performance and simplicity we now use a map for the currencyPairs.

* fix reporting when a websocket is reconnected
* fix not switching REST off after websocket available again
* fix race condition in isProcessing flag

This PR still could go further by avoiding cycling through everything each time, and by pushing some one-time work for adding enabled pairs down to a later stage off the hot path.
This was the smallest chunk of refactoring I felt could address everything without changing too much.

Significant manual testing done with a variety of Timeouts to test for edgecases and handling.

## Type of change

Please delete options that are not relevant and add an `x` in `[]` as item is complete.

- [x] Bug fix (non-breaking change which fixes an issue)

## How has this been tested

- [x] go test ./... -race
- [x] golangci-lint run
- [x] Manual testing of variety of disconnects of WS and Timeout settings

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation and regenerated documentation via the documentation tool
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally and on Github Actions/AppVeyor with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

shazbert commented on engine/sync_manager.go:
```diff
@@ -397,72 +356,72 @@ func (m *syncManager) Update(exchangeName string, p currency.Pair, a asset.Item,
+		)
+	}
+
+	return m.update(c, syncType, err)
```
With websocket enabled. `.\gocryptotrader -tickersync=true -orderbooksync=false`  Orderbooks are still being processed same for `.\gocryptotrader -tickersync=false -orderbooksync=true`. I think we can just return for now if sync type is disabled and add a TODO to limit subscriptions wrapper side.

gbjk replied:
Need to grok why that's the case. 
342-354 should have been preventing that.

gbjk replied:
Yep. I don't get why from the code.

 [342-354](https://github.com/thrasher-corp/gocryptotrader/blob/c906adfc22fa0eb77c531d5f03dd2de5464061d9/engine/sync_manager.go#L342-L354) should have guarded against that in calls to Update,
and direct internal calls to update should have been guarded by [504-512](https://github.com/thrasher-corp/gocryptotrader/blob/c906adfc22fa0eb77c531d5f03dd2de5464061d9/engine/sync_manager.go#L504-L512)

Will try to see why it's getting there.

gbjk replied:
@shazbert Do you mean this:
`[DEBUG] | SYNC | 15/06/2023 18:23:29 | Okx orderbook sync complete BTC-USDT [1/2].`
Which I still get when running:
`go run main.go -tickersync=false -orderbooksync=true`

If so, I can fix it, but it was the same on master before, so I think can be done in a separate PR.

shazbert replied:
If its on master yeah we can keep it the same. 👍 Push what you have worked on and I will send you a diff if I figure anything out. 🚀 Thanks heaps.

shazbert commented on engine/sync_manager.go:
```diff
@@ -397,72 +356,72 @@ func (m *syncManager) Update(exchangeName string, p currency.Pair, a asset.Item,
+		s.IsUsingREST = false
+		n := "Ticker"
+		if syncType == SyncItemOrderbook {
+			n = "Orderbook"
```
For REST only `.\gocryptotrader -tickersync=false -orderbooksync=true` Seems like syncing just orderbooks it alerts that ticker sync is complete. `Okx ticker sync complete LTC-USD-230623 [1/6].`

gbjk replied:
Okay. Probably not this line, since this would be giving:
` Websocket re-enabled, switching from rest to websocket`
and uppercase Ticker.
Again, will reproduce and fix or ask for help. 🙏

shazbert commented on engine/sync_manager.go:
```diff
@@ -397,72 +356,72 @@ func (m *syncManager) Update(exchangeName string, p currency.Pair, a asset.Item,
+		} else if syncType == SyncItemTrade {
+			n = "Trade"
 		}
+		log.Warnf(log.SyncMgr,
```
When disabling via cli:

```
thrasher-corp\gocryptotrader\cmd\gctcli> go run ./... websocket disable --exchange=okx
{
 "status": "success",
 "data": "websocket disabled"
}
```

Only mentions ticker `[WARN]  | SYNC | 15/06/2023 11:01:51 | Okx BTC-USDT SPOT: No ticker update after 1m0s, switching from websocket to rest` eventually all fall over to REST.

gbjk replied:
Great. Reproduced immediately. Will fix.

gbjk replied:
Whoops!
One syncBase pointer, Three trackers sharing it like a joint at a campfire 🚬 🏕️ 
Easy fix.
Probably accounts for all issues reported.

shazbert replied:
🤣 🚀

shazbert reviewed (changes requested):
Thanks for opening this PR! Looks good just a few things that are not updating on first few tries. This is much better to quickly fix issues in the package as the much larger update #1033 is very invasive.

gbjk commented:
> Thanks for opening this PR! Looks good just a few things that are not updating on first few tries. This is much better to quickly fix issues in the package as the much larger update #1033 is very invasive.

If I'd seen 1033 I'd probably not have submitted this. Glad it's useful in the short-term none-the-less 👍

gbjk commented:
Fixed. Thanks for testing and spotting 🙏

shazbert commented on engine/sync_manager.go:
```diff
@@ -397,72 +359,72 @@ func (m *syncManager) Update(exchangeName string, p currency.Pair, a asset.Item,
+}
+
+// update notifies the syncManager to change the last updated time for a exchange asset pair
+func (m *syncManager) update(c *currencyPairSyncAgent, syncType int, err error) error {
```
Running ./gocryptotrader has a diff in name tags between ticker and orderbook.
This branch:
```
[DEBUG] | SYNC | 19/06/2023 15:04:15 | Okx ticker sync complete BTC-USDT [1/12].
[INFO]  | SYNC | 19/06/2023 15:04:15 | Okx websocket BTC-USDT SPOT TICKER: Last 26429.00000000 Ask 26429.10000000 Bid 26429.00000000 High 26705.70000000 Low 26255.10000000 Volume 5387.14800189
[DEBUG] | SYNC | 19/06/2023 15:04:15 | Okx ticker sync complete WGRT-USDT [2/12].
[INFO]  | SYNC | 19/06/2023 15:04:15 | Okx websocket WGRT-USDT SPOT TICKER: Last 0.00306000 Ask 0.00309000 Bid 0.00306000 High 0.00318000 Low 0.00306000 Volume 132756.60727300
[DEBUG] | SYNC | 19/06/2023 15:04:15 | Okx ticker sync complete FSN-USDT [3/12].
[INFO]  | SYNC | 19/06/2023 15:04:15 | Okx websocket FSN-USDT SPOT TICKER: Last 0.24760000 Ask 0.24760000 Bid 0.24470000 High 0.25400000 Low 0.24290000 Volume 25332.95741200
[DEBUG] | SYNC | 19/06/2023 15:04:15 | Okx ticker sync complete LTC-USD-230623 [4/12].
[INFO]  | SYNC | 19/06/2023 15:04:15 | Okx websocket LTC-USD-230623 FUTURES TICKER: Last 77.42000000 Ask 77.33000000 Bid 77.26000000 High 78.44000000 Low 76.77000000 Volume 875.06350000
[DEBUG] | SYNC | 19/06/2023 15:04:15 | Okx ticker sync complete BTC-USD-230630-50000-C [5/12].
[INFO]  | SYNC | 19/06/2023 15:04:15 | Okx websocket BTC-USD-230630-50000-C OPTIONS TICKER: Last 0.00050000 Ask 0.00050000 Bid 0.00000000 High 0.00050000 Low 0.00050000 Volume 0.00000000
[DEBUG] | SYNC | 19/06/2023 15:04:15 | Okx ticker sync complete GRT-BTC [6/12].
[INFO]  | SYNC | 19/06/2023 15:04:15 | Okx websocket GRT-BTC MARGIN TICKER: Last 0.00000376 Ask 0.00000375 Bid 0.00000374 High 0.00000387 Low 0.00000372 Volume 5931.00000000
[INFO]  | SYNC | 19/06/2023 15:04:15 | Okx websocket GRT-BTC MARGIN TICKER: Last 0.00000376 Ask 0.00000375 Bid 0.00000374 High 0.00000387 Low 0.00000372 Volume 5931.00000000
[DEBUG] | SYNC | 19/06/2023 15:04:16 | Okx ticker sync complete WGRT-USDT [7/12].
[INFO]  | SYNC | 19/06/2023 15:04:16 | Okx websocket WGRT-USDT SPOT ORDERBOOK: Bids len: 138 Amount: 133989819.001669 WGRT. Total value: 12241.253997489193 Asks len: 361 Amount: 26946275.269770 WGRT. Total value: 21553945.33721149
[WARN]  | ORDERBOOK | 19/06/2023 15:04:16 | Potential book issue for exchange Okx pair BTC-USD-230630-50000-C asset options length Bids 0 length Asks 1
[DEBUG] | SYNC | 19/06/2023 15:04:16 | Okx ticker sync complete BTC-USDT [8/12].
[INFO]  | SYNC | 19/06/2023 15:04:16 | Okx websocket BTC-USDT SPOT ORDERBOOK: Bids len: 400 Amount: 191.025491 BTC. Total value: 5040936.126136789 Asks len: 400 Amount: 142.042803 BTC. Total value: 3757631.5236464175
[WARN]  | ORDERBOOK | 19/06/2023 15:04:16 | Potential book issue for exchange Okx pair BTC-USD-230630-50000-C asset options length Bids 0 length Asks 1
[DEBUG] | SYNC | 19/06/2023 15:04:16 | Okx ticker sync complete BTC-USD-230630-50000-C [9/12].
[INFO]  | SYNC | 19/06/2023 15:04:16 | Okx websocket BTC-USD-230630-50000-C OPTIONS ORDERBOOK: Bids len: 0 Amount: 0.000000 BTC. Total value: 0 Asks len: 1 Amount: 6560.000000 BTC. Total value: 3.2800000000000002
[DEBUG] | SYNC | 19/06/2023 15:04:16 | Okx ticker sync complete LTC-USD-230623 [10/12].
[INFO]  | SYNC | 19/06/2023 15:04:16 | Okx websocket LTC-USD-230623 FUTURES ORDERBOOK: Bids len: 35 Amount: 45463.000000 LTC. Total value: 3488379.9800000004 Asks len: 32 Amount: 47019.000000 LTC. Total value: 3659670.2099999995
[DEBUG] | SYNC | 19/06/2023 15:04:16 | Okx ticker sync complete GRT-BTC [11/12].
[INFO]  | SYNC | 19/06/2023 15:04:16 | Okx websocket FSN-USDT SPOT ORDERBOOK: Bids len: 107 Amount: 136425.007789 FSN. Total value: 21669.947007397797 Asks len: 400 Amount: 134407.786550 FSN. Total value: 69128.4497119405
[DEBUG] | SYNC | 19/06/2023 15:04:16 | Exchange CurrencyPairSyncer initial sync is complete.
[DEBUG] | SYNC | 19/06/2023 15:04:16 | Exchange CurrencyPairSyncer initial sync took 1.3246107s [12 sync items].
```

Master:
```
[DEBUG] | SYNC | 19/06/2023 15:06:12 | Okx ticker sync complete GRT-BTC [1/12].
[INFO]  | SYNC | 19/06/2023 15:06:12 | Okx websocket GRT-BTC MARGIN TICKER: Last 0.00000376 Ask 0.00000376 Bid 0.00000374 High 0.00000387 Low 0.00000372 Volume 5931.00000000
[INFO]  | SYNC | 19/06/2023 15:06:12 | Okx websocket GRT-BTC MARGIN TICKER: Last 0.00000376 Ask 0.00000376 Bid 0.00000374 High 0.00000387 Low 0.00000372 Volume 5931.00000000
[DEBUG] | SYNC | 19/06/2023 15:06:12 | Okx ticker sync complete BTC-USDT [2/12].
[INFO]  | SYNC | 19/06/2023 15:06:12 | Okx websocket BTC-USDT SPOT TICKER: Last 26428.10000000 Ask 26428.20000000 Bid 26428.10000000 High 26705.70000000 Low 26255.10000000 Volume 5384.02712319
[DEBUG] | SYNC | 19/06/2023 15:06:12 | Okx ticker sync complete WGRT-USDT [3/12].
[INFO]  | SYNC | 19/06/2023 15:06:12 | Okx websocket WGRT-USDT SPOT TICKER: Last 0.00306000 Ask 0.00309000 Bid 0.00306000 High 0.00318000 Low 0.00306000 Volume 132756.60727300
[DEBUG] | SYNC | 19/06/2023 15:06:12 | Okx ticker sync complete FSN-USDT [4/12].
[INFO]  | SYNC | 19/06/2023 15:06:12 | Okx websocket FSN-USDT SPOT TICKER: Last 0.24760000 Ask 0.24760000 Bid 0.24470000 High 0.25400000 Low 0.24290000 Volume 25332.95741200
[DEBUG] | SYNC | 19/06/2023 15:06:12 | Okx ticker sync complete BTC-USD-230630-50000-C [5/12].
[INFO]  | SYNC | 19/06/2023 15:06:12 | Okx websocket BTC-USD-230630-50000-C OPTIONS TICKER: Last 0.00050000 Ask 0.00050000 Bid 0.00000000 High 0.00050000 Low 0.00050000 Volume 0.00000000
[DEBUG] | SYNC | 19/06/2023 15:06:12 | Okx ticker sync complete LTC-USD-230623 [6/12].
[INFO]  | SYNC | 19/06/2023 15:06:12 | Okx websocket LTC-USD-230623 FUTURES TICKER: Last 77.42000000 Ask 77.35000000 Bid 77.28000000 High 78.44000000 Low 76.77000000 Volume 875.06350000
[DEBUG] | SYNC | 19/06/2023 15:06:12 | Okx orderbook sync complete WGRT-USDT [7/12].
[INFO]  | SYNC | 19/06/2023 15:06:12 | Okx websocket WGRT-USDT SPOT ORDERBOOK: Bids len: 138 Amount: 133989819.001669 WGRT. Total value: 12241.253997489193 Asks len: 361 Amount: 26946275.269770 WGRT. Total value: 21553945.33721149
[DEBUG] | SYNC | 19/06/2023 15:06:12 | Okx orderbook sync complete BTC-USDT [8/12].
[INFO]  | SYNC | 19/06/2023 15:06:12 | Okx websocket BTC-USDT SPOT ORDERBOOK: Bids len: 400 Amount: 195.743380 BTC. Total value: 5165560.187974127 Asks len: 400 Amount: 138.014803 BTC. Total value: 3651296.436886323
[DEBUG] | SYNC | 19/06/2023 15:06:12 | Okx orderbook sync complete BTC-USD-230630-50000-C [9/12].
[INFO]  | SYNC | 19/06/2023 15:06:12 | Okx websocket BTC-USD-230630-50000-C OPTIONS ORDERBOOK: Bids len: 0 Amount: 0.000000 BTC. Total value: 0 Asks len: 1 Amount: 6560.000000 BTC. Total value: 3.2800000000000002
[DEBUG] | SYNC | 19/06/2023 15:06:12 | Okx orderbook sync complete LTC-USD-230623 [10/12].
[INFO]  | SYNC | 19/06/2023 15:06:12 | Okx websocket LTC-USD-230623 FUTURES ORDERBOOK: Bids len: 32 Amount: 44234.000000 LTC. Total value: 3393916.5299999993 Asks len: 31 Amount: 45708.000000 LTC. Total value: 3558045.8899999997
[DEBUG] | SYNC | 19/06/2023 15:06:12 | Okx orderbook sync complete GRT-BTC [11/12].
[INFO]  | SYNC | 19/06/2023 15:06:12 | Okx websocket GRT-BTC MARGIN ORDERBOOK: Bids len: 27 Amount: 74502.000000 GRT. Total value: 0.23399061000000004 Asks len: 77 Amount: 218705.000000 GRT. Total value: 1.7037535199999996
[DEBUG] | SYNC | 19/06/2023 15:06:12 | Okx orderbook sync complete FSN-USDT [12/12].
[INFO]  | SYNC | 19/06/2023 15:06:12 | Okx websocket FSN-USDT SPOT ORDERBOOK: Bids len: 107 Amount: 136425.007789 FSN. Total value: 21669.947007397797 Asks len: 400 Amount: 134407.786550 FSN. Total value: 69128.4497119405
[DEBUG] | SYNC | 19/06/2023 15:06:12 | Exchange CurrencyPairSyncer initial sync is complete.
[DEBUG] | SYNC | 19/06/2023 15:06:12 | Exchange CurrencyPairSyncer initial sync took 1.3083655s [12 sync items].
```


```diff
diff --git a/engine/sync_manager.go b/engine/sync_manager.go
index 6bc4fe78b..d9bcc19d8 100644
--- a/engine/sync_manager.go
+++ b/engine/sync_manager.go
@@ -378,29 +378,30 @@ func (m *syncManager) Update(exchangeName string, p currency.Pair, a asset.Item,
        }
        s := c.trackers[syncType]

+       name := "Ticker"
+       if syncType == SyncItemOrderbook {
+               name = "Orderbook"
+       } else if syncType == SyncItemTrade {
+               name = "Trade"
+       }
+
        if !s.IsUsingWebsocket {
                s.IsUsingWebsocket = true
                s.IsUsingREST = false
-               n := "Ticker"
-               if syncType == SyncItemOrderbook {
-                       n = "Orderbook"
-               } else if syncType == SyncItemTrade {
-                       n = "Trade"
-               }
                log.Warnf(log.SyncMgr,
                        "%s %s %s: %s Websocket re-enabled, switching from rest to websocket",
                        c.Exchange,
                        m.FormatCurrency(c.Pair),
                        strings.ToUpper(c.AssetType.String()),
-                       n,
+                       name,
                )
        }

-       return m.update(c, syncType, err)
+       return m.update(c, syncType, name, err)
 }

 // update notifies the syncManager to change the last updated time for a exchange asset pair
-func (m *syncManager) update(c *currencyPairSyncAgent, syncType int, err error) error {
+func (m *syncManager) update(c *currencyPairSyncAgent, syncType int, name string, err error) error {
        if syncType < SyncItemTicker || syncType > SyncItemTrade {
                return fmt.Errorf("%v %w", syncType, errUnknownSyncItem)
        }
@@ -416,8 +417,9 @@ func (m *syncManager) update(c *currencyPairSyncAgent, syncType int, err error)
        s.IsProcessing = false
        if atomic.LoadInt32(&m.initSyncCompleted) != 1 && !origHadData {
                removedCounter++
-               log.Debugf(log.SyncMgr, "%s ticker sync complete %v [%d/%d].",
+               log.Debugf(log.SyncMgr, "%s %s sync complete %v [%d/%d].",
                        c.Exchange,
+                       name,
                        m.FormatCurrency(c.Pair),
                        removedCounter,
                        createdCounter)
@@ -588,7 +590,7 @@ func (m *syncManager) syncTicker(c *currencyPairSyncAgent, e exchange.IBotExchan
                                relayWebsocketEvent(result, "ticker_update", c.AssetType.String(), exchangeName)
                        }
                }
-               updateErr := m.update(c, SyncItemTicker, err)
+               updateErr := m.update(c, SyncItemTicker, "Ticker", err)
                if updateErr != nil {
                        log.Errorln(log.SyncMgr, updateErr)
                }
@@ -629,7 +631,7 @@ func (m *syncManager) syncOrderbook(c *currencyPairSyncAgent, e exchange.IBotExc
                                relayWebsocketEvent(result, "orderbook_update", c.AssetType.String(), e.GetName())
                        }
                }
-               updateErr := m.update(c, SyncItemOrderbook, err)
+               updateErr := m.update(c, SyncItemOrderbook, "Orderbook", err)
                if updateErr != nil {
                        log.Errorln(log.SyncMgr, updateErr)
                }
@@ -643,7 +645,7 @@ func (m *syncManager) syncTrades(c *currencyPairSyncAgent) {
        defer c.locks[SyncItemTrade].Unlock()

        if time.Since(c.trackers[SyncItemTrade].LastUpdated) > m.config.TimeoutREST {
-               err := m.update(c, SyncItemTrade, nil)
+               err := m.update(c, SyncItemTrade, "Trade", nil)
                if err != nil {
                        log.Errorln(log.SyncMgr, err)
                }
```

gbjk replied:
Thanks for the catch.
i don't think we want to have to pass around names like that.
Will just make syncType a type which implements Stringer instead, since we already have it.

shazbert reviewed (changes requested):
Thanks! Just one other thing I noticed with a difference between master. Don't have to use diff but shows the issue. 💃

shazbert reviewed (approved):
Thank you! Looks to be working well. tACK. 💃

gloriousCode commented on engine/sync_manager.go:
```diff
@@ -397,72 +361,67 @@ func (m *syncManager) Update(exchangeName string, p currency.Pair, a asset.Item,
+		s.NumErrors++
+	}
+	s.HaveData = true
+	s.IsProcessing = false
```
You mention in the PR description that you've fixed a race condition with this. Its true, but this is also the only reference to the field now, and the bool defaults to false anyway. Seems like you could remove it?

gbjk replied:
Agreed. Done. ✅

gloriousCode reviewed (changes requested):
This is a great cleanup!
`[WARN]  | SYNC | 07/07/2023 10:46:35 | Binance BTC-USDT SPOT: Ticker Websocket re-enabled, switching from rest to websocket`

I only have the most minor comments which are pretty much all optional, let me know here or via Slack if you're not going to do them and I can switch this to tACK 😄 

Found two bugs when testing this, one I detail in the comments and the other is that if we start GCT with the websocket disabled and then enable it via RPC, we can never actually switch to websocket processing because of the websocketroutine_manager. This is an issue outside of this PR

gbjk commented on engine/sync_manager.go:
```diff
@@ -700,6 +517,136 @@ func (m *syncManager) worker() {
+				if err == nil {
+					result, err = e.FetchTicker(context.TODO(), c.Pair, c.AssetType)
+				}
+				m.tickerBatchLastRequested[exchangeName] = time.Now()
```
Agreed. Break this out into an issue and I can help @Beadko with fixing it in the coming weeks.

gbjk commented on engine/sync_manager.go:
```diff
@@ -244,50 +241,54 @@ func (m *syncManager) Stop() error {
-		}
-	}
-
-	return nil, fmt.Errorf("%v %v %v %w", exchangeName, a, p, errSyncPairNotFound)
```
Well spotted. Thanks. Done. ✅

gbjk commented on engine/sync_manager.go:
```diff
@@ -312,70 +313,33 @@ func (m *syncManager) add(c *currencyPairSyncAgent) {
 // Update notifies the syncManager to change the last updated time for a exchange asset pair
-func (m *syncManager) Update(exchangeName string, p currency.Pair, a asset.Item, syncType int, err error) error {
+// And set IsUsingWebsocket to true. It should be used externally only from websocket updaters
+func (m *syncManager) Update(exchangeName string, p currency.Pair, a asset.Item, syncType syncItemType, err error) error {
```
I was avoiding changing the API during this PR, since it's public and in `iCurrencyPairSyncer`
Suggest that's a separate PR please, since I don't want to risk reverting this change as part of an unrelated change to the existing interface func.
Happy to do it immediately after though!

gloriousCode replied:
Makes sense! Happy with that

Beadko replied:
this is now done in PR #1300

gbjk commented:
All changes implemented bar one.
@gloriousCode Happy?

shazbert reviewed (changes requested):
Test failure that needs to be rectified. Then this looks good to go. Thanks.

```
--- FAIL: TestSyncManagerUpdate (0.00s)
    sync_manager_test.go:321: received could not sync new data for    Orderbook, but expected: <nil>
Exchange currency pair syncer config: continuous: true ticker: false orderbook: false trades: true workers: 15 verbose: false 
```

gbjk commented:
@shazbert Sorry about the test.
I had a regression using kraken, fixed it but didn't see the test failure.

Hopefully good now.
No changes to implementation here, just fixed the test.

gloriousCode reviewed (approved):
tACK!

shazbert reviewed (approved):
tACK, fine work!

thrasher- reviewed (approved):
Great work @gbjk, love the simplification and improvements along the way
